docs/developer/architecture/user_customization_and_persistence.rst

```diff
@@ -0,0 +1,214 @@
+User Customization & Persistence
+================================
+
+Context
+-------
+
+.. todo::
+
+   This section will be handled by the requirements PR.
+
+
+Design Choices
+--------------
+
+:term:`ConfigMap` KV(Key-Value) store is chosen as a unified data access and
+storage media for user editable configurations in a MetalK8s cluster based on
+the above requirements for the following reasons:
+
+* The limited number of read/write operations even for multi-node setups means
+  that concurrency issues on CRUD operations will not lag even if we do not
+  implement cross-app locking.
+* Ability to support CRUD operations on ConfigMap's with CLI and UI easily
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly(modular design nature)
+* Ease of operation and a guarantee of no downtime in a MetalK8s cluster
+  especially when configuration changes are made on the fly.
+
+Rejected design choices
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Consul KV store
+~~~~~~~~~~~~~~~
+
+This approach offers a full fledge KV store with a /kv endpoint which allows
+CRUD operations on all KV data stored in it.
+Consul KV also allows access to past versions of objects and has an optimistic
+concurrency(CRUD operations will not lag with load).
+
+Note that, Consul KV store was rejected because managing operations such as
+performing full backups, system restores for a full fledged KV system
+requires time and much more efforts than the ConfigMap KV store which is
+simplistic and matches the requirements stated.
+
+
+Implementation Details
+----------------------
+
+A sample ConfigMap KV store can be defined with the following fields. These
+objects will be namespace bounded ensuring that sensitive information is
+restricted from a cluster-wide access.
+
+An example of such a ConfigMap KV store:
+
+   .. code-block:: yaml
+
+       apiVersion: v1
+       kind: ConfigMap
+       metadata:
+         namespace: <metalk8s-namespace>
+         name: configmap-kv
+       data:
+         <key>: <value>
+         <key>: <value>
+         <key>: <value>
+
+**Use case 1:**
+
+Configure and store the number of replicas for all Deployments found in the
+`metalk8s-monitoring` namespace using the ConfigMap KV store format.
+
+   .. code-block:: yaml
+
+       apiVersion: v1
+       kind: ConfigMap
+       metadata:
+         namespace: metalk8s-monitoring
+         name: <example-name>
+       data:
+         replicas.grafana: 2
+         replicas.prometheus: 2
+         replicas.alertmanager: 1
```

Comment:
This kind of format is not straightforward to manage with code, since you'll need to parse the flat keys into hierarchical structures. As discussed during the short design session, we'd rather have something more "K8s-like":

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: metalk8s-monitoring
  name: metalk8s-grafana-userconfig
data:
  config.yaml: |-
    apiVersion: metalk8s.scality.com/v1alpha1
    kind: GrafanaUserConfig
    spec:
      replicas: 2
```

This also has the benefit of bearing a schema version, which we can check before inferring the underlying service configuration.

Reply:
Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.

What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

Reply:
Not really agree with your approach here.
If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.
Also IMO we should do one configmap per "object/service" (e.g.: One configmap for auth/dex, one for grafana, one for prometheus (or maybe one for monitoring regrouping prom, grafana altermanager), one for ...)

Reply:
> Thanks for the reminder! even though i think at the end of the day, we would have littered the entire namespace with many small ConfigMap objects.
> 
> What are the chances that we could group the number of replicas specific ConfigMap for all services in the Monitoring namespace?

Oh, let us keep it service-specific as suggested since the configmap of one particular service might grow drastically when new customizations are included

Reply:
> If we use a config map IMO we should have just "key value" like Claude did and when we will use specific CRD then we could have object like you suggest.

Then it means we'll need to implement the parsing logic twice? Why not stick with the same YAML style/format?

Reply:
The configmap will be read from salt so if we use the "kubernetes style" that mean we need to implement it in our salt modules as we will have a configmap with several "key=string" and we will need to then load and parse the string to get the yaml we want.
If we stick with key=value we will just have to read a classic dict in salt so it's really easy.